### PR TITLE
SAMZA-1817: Adding pathing jar support for Long Classpath

### PR DESCRIPTION
**Pull request by Sanil15**
- To support long classpath with deterministic jar loading we use a concept called pathing jar 
- A new archive called pathing.jar is created with a custom manifest
- This custom manifest file contains the deterministic classpath that we construct

Readings:
- [Adding Classes to the JAR File's Classpath  
](https://docs.oracle.com/javase/tutorial/deployment/jar/downman.html)
- [How to avoid argument line too long with manifest files
](https://stackoverflow.com/questions/3057841/too-long-line-in-manifest-file-while-trying-to-create-jar)
- [Pathing Jar information: (from tools)
](https://stackoverflow.com/questions/201816/how-to-set-a-long-java-classpath-in-windows)

### CODE REVIEW

**Comment by Sanil15** [comment]
@abhishekshivanna @sborya Please take a look

File: samza-shell/src/main/bash/run-class.sh [hunk 1]
```diff
@@ -62,26 +62,30 @@ if [ -d "$JOB_LIB_DIR" ] && [ "$JOB_LIB_DIR" != "$BASE_LIB_DIR" ]; then
...
-  for file in $BASE_LIB_DIR/*.war;
+  # default behavior, all the jars need to be appended on newlines
+  # to ensure line argument length of 72 bytes is not violated
+  for file in $BASE_LIB_DIR/*.[jw]ar;
   do
-    CLASSPATH=$CLASSPATH:$file
+    CLASSPATH=$CLASSPATH"\n $file "
   done
-  echo generated from BASE_LIB_DIR CLASSPATH=$CLASSPATH
+  echo generated combined CLASSPATH=$CLASSPATH
```
Comment (sborya) [inline samza-shell/src/main/bash/run-class.sh]: this line and line #73 look the same. Please leave 'from BASE_LIB_DIR' so we can tell which code path was being run.

File: samza-shell/src/main/bash/run-class.sh [hunk 2]
```diff
@@ -62,26 +62,30 @@ if [ -d "$JOB_LIB_DIR" ] && [ "$JOB_LIB_DIR" != "$BASE_LIB_DIR" ]; then
   job_jars=`for file in $JOB_LIB_DIR/*.[jw]ar; do name=\`basename $file\`; if [[ $base_jars != *"$name"* ]]; then echo "$file"; fi; done`
   # get all lib jars and reverse sort it by versions
   all_jars=`for file in $base_jars $job_jars; do echo \`basename $file|sed 's/.*[-]\([0-9]\+\..*\)[jw]ar$/\1/'\` $file; done|sort -t. -k 1,1nr -k 2,2nr -k 3,3nr -k 4,4nr|awk '{print $2}'`
-  # generate the class path based on the sorted result
-  for jar in $all_jars; do CLASSPATH=$CLASSPATH:$jar; done
+  # generate the class path based on the sorted result, all the jars need to be appended on newlines
+  # to ensure java argument length of 72 bytes is not violated
```
[the remark this answers is not in the document]
Comment (Sanil15) [reply samza-shell/src/main/bash/run-class.sh]: This fails with line length too long/

how to ge through this?
Relative Paths :)

[the remark this answers is not in the document]
Comment (Sanil15) [reply samza-shell/src/main/bash/run-class.sh]: This is from Jar [Spec](https://docs.oracle.com/javase/8/docs/technotes/guides/jar/jar.html#Notes_on_Manifest_and_Signature_Files
): 

No line may be longer than 72 bytes (not characters), in its UTF8-encoded form. If a value would make the initial line longer than this, it should be continued on extra lines (each starting with a single SPACE).

[the remark this answers is not in the document]
Comment (sborya) [reply samza-shell/src/main/bash/run-class.sh]: great

File: samza-shell/src/main/bash/run-class.sh [hunk 3]
```diff
@@ -62,26 +62,30 @@ if [ -d "$JOB_LIB_DIR" ] && [ "$JOB_LIB_DIR" != "$BASE_LIB_DIR" ]; then
...
-    CLASSPATH=$CLASSPATH:$file
+    CLASSPATH=$CLASSPATH"\n $file "
   done
   echo generated from BASE_LIB_DIR CLASSPATH=$CLASSPATH
 fi
 
+# Newlines and spaces are intended to ensure proper parsing of manifest in pathing jar
+printf "Class-Path: \n $CLASSPATH \n\n" > manifest.txt
+# Creates a new archive and adds custom manifest information to pathing.jar
+jar -cvmf manifest.txt pathing.jar
```
Comment (shanthoosh) [inline samza-shell/src/main/bash/run-class.sh]: Would be better to create a constant JAR and set it to $JAVA_HOME/bin/jar if JAVA_HOME is defined.

```bash
if [ -z "$JAVA_HOME" ]; then
  JAR="jar"
else
  JAR="$JAVA_HOME/bin/jar"
fi

```

Comment (Sanil15) [reply samza-shell/src/main/bash/run-class.sh]: Yup I already identified this while testing. Will do

**Comment by vjagadish1989** [comment]
@shanthoosh is this good to merge or do you still have unresolved questions on this PR?

**Comment by shanthoosh** [comment]
@vjagadish1989 
I am not sure of the status of testing. Sanil might have more information about it. 
There's still one unresolved comment, which was agreed to be changed in the above comment.

**Comment by abhishekshivanna** [comment]
LGTM

**Pull request closed without merging**